services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/grid/internal/replica/validation/GridReplicaValidationManagerImpl.java

```diff
@@ -0,0 +1,169 @@
+		this.patchBuilderPublisher = patchBuilderPublisher;
+	}
+
+	@Override
+	public void validateChanges(String sessionId, Long replicaId, String connectionId,
+			Collection<LogicalTimestamp> changedVectorIds) {
+		ValidateArgument.required(sessionId, "sessionId");
+		ValidateArgument.required(replicaId, "replicaId");
+		ValidateArgument.required(connectionId, "connectionId");
+
+		if (changedVectorIds == null || changedVectorIds.isEmpty()) {
+			return;
+		}
+
+		Optional<GridSession> gridSession = gridDao.getGridSession(sessionId);
+		if (!hasValidSession(gridSession)) {
+			return;
+		}
+
+		Optional<GridHeader> header = gridReplicaViewManager.readHeader(sessionId, replicaId);
+		if (header.isEmpty()) {
+			return;
+		}
+
+		List<RowView> rowsToValidate = getRowsToValidate(header.get(), changedVectorIds);
+		if (rowsToValidate.isEmpty()) {
+			return;
+		}
+
+		List<IntendedChange> intendedChanges = validateRows(header.get(), gridSession.get().getGridJsonSchema$Id(),
+				rowsToValidate);
+		if (intendedChanges.isEmpty()) {
+			return;
+		}
+
+		// send the changes to the patch builder.
+		patchBuilderPublisher.sendChangesToPatchBuilder(new IntendedChangeSet().setChanges(intendedChanges)
+				.setSessionId(sessionId).setReplicaId(replicaId).setConnectionId(connectionId));
+	}
+
+	boolean hasValidSession(Optional<GridSession> gridSession) {
+		return gridSession.isPresent() && gridSession.get().getGridJsonSchema$Id() != null;
+	}
+
+	/**
+	 * Run a query to fetch the rows identified by the provided vector IDs.
+	 * 
+	 * @param header
+	 * @param changedVectorIds
+	 * @return
+	 */
+	List<RowView> getRowsToValidate(GridHeader header, Collection<LogicalTimestamp> changedVectorIds) {
+		List<LogicalTimestamp> vectorList = changedVectorIds.stream().collect(Collectors.toList());
+		Long limit = (long) (changedVectorIds.size() + 1);
+		return gridReplicaViewManager.querySinglePage(header, List.of(new VectorIdViewFilter(vectorList)), limit, 0L);
+	}
+
+	/**
+	 * Validate a set of rows and generate changes to be sent to the patch builder.
+	 * 
+	 * @param header
+	 * @param schemaId
+	 * @param rowsToValidate
+	 * @return
+	 */
+	List<IntendedChange> validateRows(GridHeader header, String schemaId, List<RowView> rowsToValidate) {
+		JsonSchema schema = jsonSchemaManager.getValidationSchema(schemaId);
+
+		return rowsToValidate.stream().map(row -> validateCells(header, schema, row)).filter(Optional::isPresent)
+				.map(Optional::get).collect(Collectors.toList());
+	}
+
+	/**
+	 * Validate the cell data for a single row against the provided schema.
+	 * 
+	 * @param header
+	 * @param schema
+	 * @param row
+	 * @return Optional.empty() if the new results are equal to the old.
+	 */
+	Optional<IntendedChange> validateCells(GridHeader header, JsonSchema schema, RowView row) {
+		RowJsonSubject subject = new RowJsonSubject(header.getOrderedColumns(), row);
+		ValidationResults validationResults = jsonSchemaValidationManager.validate(schema, subject);
```

Comment:
`validate` builds the schema each time it is called. Change the `jsonSchemaValidationManager` implementation to accept a list of subjects to validate.